### ISSUE TITLE
sdpa for bert causes nan output when mix-precision enabled.

### ISSUE THREAD

**Issue by Leoyzen** [huggingface/transformers#31038]
### System Info

* transformers: transformers==4.41.1
* pytorch: 2.3.0+cu121
* python: 3.10

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

training bert model with f16-mixed /bf16-mixed enabled.

```python
# modeling_bert.py
...
        # Expand the attention mask
        if use_sdpa_attention_masks:
            # Expand the attention mask for SDPA.
            # [bsz, seq_len] -> [bsz, 1, seq_len, seq_len]
            if self.config.is_decoder:
                extended_attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
                    attention_mask,
                    input_shape,
                    embedding_output,
                    past_key_values_length,
                )
            else:
                extended_attention_mask = _prepare_4d_attention_mask_for_sdpa(
                    attention_mask, embedding_output.dtype, tgt_len=seq_length
                )
            #  print(f"sdpa attn mask:{embedding_output.dtype=} {extended_attention_mask.sum(axis=(2,3))=}")
        else:
            # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
            # ourselves in which case we just need to make it broadcastable to all heads.
            extended_attention_mask = self.get_extended_attention_mask(attention_mask, input_shape)
...
```
In the code above, the `embedding_output.dtype` used to produce the attention mask.

But within mixed precision training, the dtype of embedding output should always be float32, which casues the attention mask to be float32, then causes `F.scaled_dot_product_attention` failed and outputs nan.

We should specify the dtype by another way instead of using dtype of embedding outputs.

see also https://github.com/pytorch/pytorch/issues/98814#issuecomment-1796144712

### Expected behavior

training with f16-mixed/bf16-mixed mixed precision should ok and without nan output.

**Comment by amyeroberts** [comment]
cc @ArthurZucker @younesbelkada

**Comment by ArthurZucker** [comment]
do you want to open a PR for the fix? 🤗 this sounds indeed like a problem that we can avoid\

**Comment by Leoyzen** [comment]
@ArthurZucker I'm not sure which dtype should been used for attention_mask generation. Do you have a recommendation?
